### PR TITLE
GMP Pruning and Related Functionality

### PR DESCRIPTION
**Pull request by mvacaporale**
This PR is for GMP pruning mixins, configs, and related functionality. It includes the following changes:
- Enable fully dense but sparsifiable models at initialization 
- Updated wide tiny-bert experiments
- Add script to create a prunable checkpoint (one with SparseWeight modules) of a densely trained model.
- LR schedule for GMP pruning on a fully dense model.
- The `RezeroWeights` callback is now configurable to log every `log_steps`  instead of after every training step.
- New mixins: `GradualMagnitudePruningMixin` and  `ThreeStageLRMixin`

The last two mixins can be used together or independently depending on the use case. One may apply GMP pruning during pre-training or afterwards. If the latter, `ThreeStageLRMixin` should be used to enable lr phases of stabilization, pruning, and fine-tuning. Otherwise, a OneCycle LR or other schedule may be used for GMP throughout pre-training. As of now, there are experiments that try out both methods and it's an open question as to which leads to the best eval-loss.

### CODE REVIEW

File: nupic/research/frameworks/dynamic_sparse/gmp_lr_scheduler.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+        # Phase 2: Pruning phase - constant lr.
+        else:
+            return self.pruning_lr(step)
+
+    # Warm-up phase.
+    def warmup_lr(self, step: int):
+        """Linearly ramp up the lr from 0 to max_lr in `warmup_steps`"""
+        min_lr = 0
+        return (self.max_lr - min_lr) / (self.warmup_steps - 1) * step + min_lr
```
Comment (benja-matic) [inline nupic/research/frameworks/dynamic_sparse/gmp_lr_scheduler.py:92]: Why set min_lr to zero every time? If min_lr is 0, why have + or - min_lr on line 92? If there is ever a case where min_lr is not 0, maybe set that as an attribute? Also, not crucial, but it looks to me like the first part of that expression `(self.max_lr - min_lr) / (self.warmup_steps - 1)` is just an increment rate, which can be computed once in advance. Then you could do something like 
`self.warmup_rate = (self.max_lr - self.min_lr) / (self.warmup_steps - 1)`
in the init, followed by
`return self.min_lr + (self.warmup_rate * step)`
in the warmup_lr function.

File: nupic/research/frameworks/dynamic_sparse/gmp_lr_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        return self.max_lr
+
+    # Cool-down phase
+    def cooldown_lr(self, step: int):
+        """
+        Similar to StepLR, decay the learning twice throughout the cooldown phase.
+        Right now it only supports decaying the lr twice.
+        """
+
+        step_size = int(self.cooldown_steps / 3)
```
Comment (benja-matic) [inline nupic/research/frameworks/dynamic_sparse/gmp_lr_scheduler.py:106]: Similarly, you could compute the cooldown rate once in advance, though it's not a big difference.

**Review by benja-matic — APPROVED** [review]
Very nice. These new mixins seem like great tools to have in our kit. I left a couple comments but LGTM.

File: nupic/research/frameworks/dynamic_sparse/gmp_lr_scheduler.py (hunk 1 above)
Comment (lucasosouza) [reply nupic/research/frameworks/dynamic_sparse/gmp_lr_scheduler.py:92]: Similar comment, maybe min_lr should be a class attribute that can be defined by the user.

**Review by lucasosouza — APPROVED** [review]
I did a quick review as Ben had reviewed already. What is the difference between the 3 stage LR scheduler and one cycle, and which one should we use and when?

File: projects/transformers/trainer_mixins/gmp.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+            cooldown_gamma=self.cooldown_gamma,
+        )
+
+
+def get_model(model):
+    """Extract model from DistributedDataParallel."""
+    if isinstance(model, DistributedDataParallel):
+        return model.module
+    else:
+        return model
```
Comment (lscheinkman) [inline projects/transformers/trainer_mixins/gmp.py:285]: The model may be wrapped with deepspeed. It is safer to use HF version of this method:
```
from transformers.modeling_utils import unwrap_model
model = unwrap_model(model)
```
```
